### DISCUSSION TITLE
Would you like to have also Kotlin extensions for LangChain4j?

### DISCUSSION THREAD

**Discussion by kpavlov** [langchain4j/langchain4j#1897] [Ideas]
The code can be organized similar to how it is done in SpringBoot with sourceRoot for Kotlin stuff: https://github.com/spring-projects/spring-boot/tree/main/spring-boot-project/spring-boot/src/main

**Comment by glaforge** [comment]
Just so I understand the scope of this ticket: we're not adding a mandatory dependency on Kotlin in LangChain4j, right?
If we're adding some convenience for other JJVM languages, it shouldn't be imposed onto others who are just using Java.
I don't want to have a dependency on Kotlin if I'm not using it in my project.

**Reply by kpavlov** [reply]
In practice, there is a big chance, that there is already a runtime dependency on Kotlin stdlib, e.g. in langchain4j-google-ai-gemini 😄 
```
[INFO] +- com.squareup.okhttp3:logging-interceptor:jar:4.12.0:compile
[INFO] |  \- org.jetbrains.kotlin:kotlin-stdlib-jdk8:jar:1.9.10:compile
[INFO] |     +- org.jetbrains.kotlin:kotlin-stdlib:jar:1.9.10:compile
[INFO] |     |  \- org.jetbrains:annotations:jar:13.0:compile
[INFO] |     \- org.jetbrains.kotlin:kotlin-stdlib-jdk7:jar:1.9.10:compile
```
But no, Kotlin should not be a requirement. kotlin-stdlib might be an `optional` maven dependency, in this case kotlin sources can live in the same module. Or it can be a separate artifact: module or repo

**Reply by glaforge** [reply]
Retrofit/OkHttp uses Kotlin, but there's an ongoing PR to just use HttpClient from the JDK, so Kotlin depds will disappear soon from that module.

**Reply by kpavlov** [reply]
> which http client is most commonly used in Kotlin?

Kotlin is JVM language, so any java client can be used. But for multiplatform code there is a [KTor client](https://ktor.io/docs/client-create-new-application.html) which is actually a wrapper around IO libraries like Netty

**Reply by kpavlov** [reply]
> Retrofit/OkHttp uses Kotlin, but there's an ongoing PR to just use HttpClient from the JDK, so Kotlin depds will disappear soon from that module.

I am not suggesting to rewrite the whole library to Kotlin or replace IO layer, even if Kotlin implementation would have been more efficient (with coroutines) and, potentially, multiplatform :) I suggest adding a wrappers to make using LC4J API from Kotlin more easy and natural.

Actually, already started: https://github.com/kpavlov/langchain4j-kotlin

**Comment by kpavlov** [comment]
This is the [example](https://github.com/kpavlov/langchain4j-kotlin/blob/main/langchain4j-core-kotlin/src/test/kotlin/me/kpavlov/langchain4j/kotlin/StreamingChatLanguageModelIT.kt#L58-L70) how to use [`Flows`](https://kotlinlang.org/docs/flow.html) with `StreamingChatLanguageModel`:

```kotlin
val collectedTokens = mutableListOf<String>()

model.generateFlow(messages)
    .collect {
        logger.info("Received event: $it")
        when (it) {
            is Token -> {
                logger.info("Token: '${it.token}'")
                collectedTokens.add(it.token)
            }

            is Completion -> responseRef.set(it.response)
            else -> fail("Unsupported event: $it")
        }
    }
```

The extension function looks like [this](https://github.com/kpavlov/langchain4j-kotlin/blob/main/langchain4j-core-kotlin/src/main/kotlin/dev/langchain4j/model/chat/StreamingChatLanguageModelExtensions.kt#L65-L96):
```kotlin
fun StreamingChatLanguageModel.generateFlow(
    messages: List<ChatMessage>,
): Flow<StreamingChatLanguageModelReply> = callbackFlow {

    val model = this@generateFlow

    val handler = object : StreamingResponseHandler<AiMessage> {
        override fun onNext(token: String) {
            logger.trace(PII, "Received token: {}", token)
            trySend(StreamingChatLanguageModelReply.Token(token))
        }

        override fun onComplete(response: Response<AiMessage>) {
            logger.trace(PII, "Received response: {}", response)
            trySend(StreamingChatLanguageModelReply.Completion(response))
            close()
        }

        override fun onError(error: Throwable) {
            close(error)
        }
    }

    logger.info("Starting flow...")
    model.generate(messages, handler)

    // This will be called when the flow collection is closed or cancelled.
    awaitClose {
        // cleanup
        logger.info("Flow is canceled")
    }
}
```

**Comment by kpavlov** [comment]
Porting [kpavlov/langchain4j-kotlin](https://github.com/kpavlov/langchain4j-kotlin)

Integrated Kotlin support by adding necessary dependencies and plugins to the project. Introduced Kotlin extensions for ChatLanguageModel with corresponding tests. Updated .editorconfig for improved import and properties formatting.

- Kotlin dependencies added as optional, so it will not pollute java projects
 - Maven Javadoc plugin replaced with [Dokka](https://kotlinlang.org/docs/dokka-maven.html) in core module in order to include both Java and Kotlin docs
    
PR #2197
